extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/extractors/SoundcloudStreamExtractor.java

```diff
@@ -190,50 +192,109 @@ public String getHlsUrl() {
         try {
             final JsonArray transcodings = track.getObject("media").getArray("transcodings");
 
-            // get information about what stream formats are available
-            for (Object transcoding : transcodings) {
+            // Iterate a first time to see if there is a progressive MP3 stream available.
+            // If yes, the MP3 HLS stream will be not added to audioStreams.
 
+            boolean mp3ProgressiveStreamInTranscodings = false;
+
+            for (final Object transcoding : transcodings) {
+                final JsonObject t = (JsonObject) transcoding;
+                if (t.getString("preset").contains("mp3") &&
+                    t.getObject("format").getString("protocol").equals("progressive")) {
+                    mp3ProgressiveStreamInTranscodings = true;
+                }
+            }
+
+            // Get information about what stream formats are available
+            for (final Object transcoding : transcodings) {
                 final JsonObject t = (JsonObject) transcoding;
                 String url = t.getString("url");
+                final String mediaUrl;
+                final MediaFormat mediaFormat;
+                final int bitrate;
 
                 if (!isNullOrEmpty(url)) {
+                    if (t.getString("preset").contains("mp3")) {
+                        // Don't add the MP3 HLS stream if there is a progressive stream present
+                        // because the two have the same bitrate
+                        if (t.getObject("format").getString("protocol").equals("hls") &&
+                            mp3ProgressiveStreamInTranscodings) {
+                            continue;
+                        }
 
-                    // We can only play the mp3 format, but not handle m3u playlists / streams.
-                    // what about Opus?
-                    if (t.getString("preset").contains("mp3")
-                            && t.getObject("format").getString("protocol").equals("progressive")) {
+                        mediaFormat = MediaFormat.MP3;
+                        bitrate = 128;
+                    } else if (t.getString("preset").contains("opus")) {
+                        mediaFormat = MediaFormat.OPUS;
+                        bitrate = 64;
+                    } else {
+                        continue;
+                    }
+
+                    // TODO: move this to a separate method to generate valid urls when needed (e.g. resuming a paused stream)
```

Comment:
todo

Reply:
that's a todo i left while hotfixing soundcloud. It should be done, but in a separate PR.